system_metrics_collector/src/topic_statistics_collector/received_message_age.hpp

```diff
@@ -56,41 +80,18 @@ class ReceivedMessageAgeCollector : public TopicStatisticsCollector<T>
   *
   * @param received_message, the message to calculate age of.
   */
-  void OnMessageReceived(const T & received_message) override
+  void OnMessageReceived(
+    const T & received_message,
+    const rcl_time_point_value_t & now_nanoseconds) override
   {
-    const auto timestamp_from_header = message_filters::message_traits::TimeStamp<T>::value(
-      received_message);
-
-    if (timestamp_from_header.nanoseconds()) {
-      const auto now = GetCurrentTime();
-
-      if (timestamp_from_header.get_clock_type() == now.get_clock_type()) {
-        const std::chrono::nanoseconds age_nanos{
-          now.nanoseconds() - timestamp_from_header.nanoseconds()};
-        const auto age_millis = std::chrono::duration_cast<std::chrono::milliseconds>(age_nanos);
-
-        system_metrics_collector::Collector::AcceptData(static_cast<double>(age_millis.count()));
-      } else {
-        std::stringstream warn_msg;
-        warn_msg <<
-          "Message header and current clock have different time sources, " <<
-          "cannot measure message age" <<
-          "\nmessage header clock type: " << timestamp_from_header.get_clock_type() <<
-          "\ncollector clock type: " << clock_.get_clock_type();
-        RCUTILS_LOG_WARN_NAMED("topic_statistics_collector", "%s", warn_msg.str().c_str());
-      }
-    }
-  }
+    const auto timestamp_from_header = TimeStamp<T>::value(received_message);
```

Comment:
What is this template magic trying to do?

Reply:
Is this what the circular dependency involving `message_filters` was needed for? And I am guessing that [this](https://github.com/ros-tooling/system_metrics_collector/pull/113/files#diff-5d4109d4db4bc0aeccb452eada7901beR29-R56) is copied and pasted from `message_filters` to break that circular dependency? Is there some way to additionally simplify this?

Reply:
> What is this template magic trying to do?

Looks like the logic here is if message type `T` has a field called `header`, then the timestamp will be taken from that `header`, otherwise the timestamp will simply be 0. And later, if `timestamp_from_header == 0`, you won't accept the data.

Reply:
Yes, you are correct. If a header is present we attempt to extract a valid value. Notice how this doesn't return rclcpp::Time, but the rcl type (unlike [message_filters](https://github.com/ros2/message_filters/blob/master/include/message_filters/message_traits.h#L89)).

Reply:
I think it's better to refactor `TimeStamp::value()` into standalone templated function(s), move those function into this class as private helper function(s) (looking like [this](https://stackoverflow.com/a/13402245)), move the `AcceptData()` logic into this private helper function(s). Finally, have `OnMessageReceived()` call this private helper function. (`OnMessageReceived()` can't become templated function(s) itself because it's virtual.)

Reply:
I'm not seeing the advantage, especially since it seems strange to branch the header extraction compared to doing it by example.

Reply:
My suggestion also brings up another question. What if message `T` has a field called `header`, but the timestamp is actually 0? Who is setting these headers? Can everybody be trusted to not start (simulation) time at 0?

Reply:
Good point, if you feel it's worth it, we could return `std::pair(int64_t, bool /* is_initialized */)` which is an OK alternative to `std::optional` we can't use yet.